Random Forest Models

jds485 opened the pull request:
This is a large PR that provides feature selection, RF models, and plot diagnostics for several model iterations. I'm suggesting 2 separate reviews that focus on 1) feature selection and RF models, and 2) plots. Hopefully that makes it more manageable to review. 

I don't think you need to try running these targets because they take a while to complete. If you want to inspect the completed targets, you can load the object targets from S3 and check out the file targets in the Caldera directory (I can send you a link to these directories outside of GitHub). Let me know if you need help finding any results.


**Galen Review:** RF models (targets in `4_predict.R`) and function files:
- `4_predict/src/select_features.R` - functions for feature selection used in target `p4_screened_attrs`
- `4_predict/src/train_models.R` - functions for remaining feature selection and RF modeling targets in `4_predict.R`. 


**Lauren Review:** Plots (targets in `4_predict_plots.R`) and function files:
- `3_visualize/src/plot_nhdv2_attr.R` - edits to add the full network even if a reach has no data for an attribute (important for model predictions, not for our gathered attributes).
- `_targets.R` - Question: there are now many directories specified. Maybe there is a cleaner way to create these? The recursive arg. could work. Or maybe for the train/test splits, a function with a keyword ("random" or "temporal" in this case).
- `4_predict/src/plot_diagnostics.R` - many new plotting functions here for the prediction models. I'd suggest looking at the figures first to get an idea of what they look like and then reviewing the plot functions / targets.
- There are a lot of similar targets for each of the unique train/test + attribute iterations. If you see a good way to implement targets branching over these iterations to make the code shorter / cleaner, let me know. I did not give much thought to that while coding it, but realized when making this PR that it could be useful, especially if more model iterations are added.


**Not in need of review:**
- The `1_fetch/src/generate_credentials.R` function was reviewed in #181. It's showing as new here because of how the merge was handled.
- slurm script. This was updated to correct a quotation mark typo.
- targets meta file

lekoenig commented on _targets.R:
```diff
@@ -35,6 +39,103 @@ dir.create("3_visualize/out/daily_timeseries_png/",showWarnings = FALSE)
 dir.create("3_visualize/out/hourly_timeseries_png/",showWarnings = FALSE)
 dir.create("3_visualize/out/nhdv2_attr_png/",showWarnings = FALSE)
 dir.create("3_visualize/out/nhdv2_attr_png/refined",showWarnings = FALSE)
+dir.create("4_predict/out/",showWarnings = FALSE)
```
Yeah, that is a lot of new directories 😄. When I create new pipelines, I've actually started taking a different approach where I commit blank text files called `placeholder.txt` so that a user doesn't need to create new directories when they clone a repo, and the `_targets.R` file doesn't get so cluttered. This works well for select, "primary" directories like "2_process/out", but it's probably not practical for creating all of these train/test directories.

At minimum, I think we could cut down on a few of these lines by using `recursive = TRUE` as you mention in your overview comments. To collapse this further and define many new directories at once, my first approach would be to use `expand.grid()` to come up with the different combinations. I tried this briefly and it seems to work (although I have not carefully checked that I've captured all of the directories you have below). The step to define `file_path` using the `mutate` step feels a bit hacky and isn't very scalable, but I ran out of time to play with it for now.

```r
# create directories for each of the following combinations of data splits,
# feature sets and temporal resolutions:
predict_dir <- "4_predict/out"
train_test_other <- c("vip","hypopt") # wasn't sure what hypopt refers to, so just call this vector "other" for now
train_test_splits <- c("temporal","random")
train_test_features <- c("RF_static_dynamic", "RF_min_static_dynamic","RF_dynamic")
train_test_res <- c("pred_obs","spatial_res","monthly_res","annual_res","temporal_res")
rf_xai_plot_types <- c("shap","dependence") 
rf_xai_dep_options <- c("pdp","ice")
train_test_dir <- bind_rows(
  expand.grid(predict_dir, train_test_splits, train_test_res, train_test_features),
  expand.grid(predict_dir, train_test_splits, train_test_other),
  expand.grid(predict_dir, train_test_splits, rf_xai_plot_types[rf_xai_plot_types == "shap"], train_test_features),
  expand.grid(predict_dir, train_test_splits, rf_xai_plot_types[rf_xai_plot_types == "dependence"], train_test_features, rf_xai_dep_options)) %>%
  mutate(file_path = case_when(is.na(Var4) ~ file.path(Var1, Var2, Var3),
                               is.na(Var5) ~ file.path(Var1, Var2, Var3, Var4),
                               TRUE ~ file.path(Var1, Var2, Var3, Var4, Var5))) %>%
  pull(file_path)

# Now actually create the newly-defined directories. 
# You could also use sapply(train_test_dir, dir.create, recursive = TRUE) 
# but I just used a for loop since all we need is to evaluate the expression, 
# rather than return any output.
for(i in seq_along(train_test_dir)){
  dir.create(train_test_dir[i], recursive = TRUE)
}
```

This is more compact, but is also less readable than manually defining each of the new directories on a new line. What do you think?

jds485 replied:
Let me see if what you wrote here generates all the directories. And I can see if the mutate step can be simplified

jds485 replied:
I added 2 items to `train_test_features` that are needed for the `random` split. That generates unnecessary directories for the `temporal` split, but I'm okay with that to have a simpler directory generation method.

lekoenig commented on 4_predict_plots.R:
```diff
@@ -0,0 +1,1845 @@
+  tar_target(p4_vip_static_png,
+             plot_vip(RF_model = p4_train_RF_static$best_fit,
+                      model_name = 'daily_SC_RF_static',
+                      num_features = 20,
```
I like this option to set how many features should be included in the vip plot.

lekoenig commented on 4_predict_plots.R:
```diff
@@ -0,0 +1,1845 @@
+             format = 'file',
+             repository = 'local'
+  ),
+  tar_target(p4_vip_static_dynamic_png,
+             plot_vip(RF_model = p4_train_RF_static_dynamic$best_fit,
+                      model_name = 'daily_SC_RF_static_dynamic',
+                      num_features = 20,
+                      out_dir = '4_predict/out/random/vip'),
+             deployment = 'main',
+             format = 'file',
+             repository = 'local'
+  ),
```
It does look like a lot of these targets are repeated across different versions of the RF models and so might be good candidates for leveraging targets' branching capabilities. 

`targets` requires that we map over another target (as opposed to some vector or data frame in the global environment), so we would need to add a target that specifies the model versions. One option would be to compile the different models within a list such that `plot_vip` could be mapped over each element within that list. 

What's not as clear to me without playing with the code is how we would define the `out_dir` for each of those branches. I can try to think more about it, and maybe adding branching in `4_predict_plots.R` is something we could add as an enhancement issue that gets tackled apart from the rest of this PR.

lekoenig replied:
What is the `str()` of `p4_train_RF_static_dynamic`?

jds485 replied:
> adding branching in 4_predict_plots.R is something we could add as an enhancement issue that gets tackled apart from the rest of this PR

I agree this is more an enhancement. I can add an issue

jds485 replied:
`p4_train_RF_static_dynamic` is a list that contains a table of model results, the best fit model evaluated on the test dataset, and the model workflow that can be used to predict with a new dataset (e.g., within `predict()`)

lekoenig replied:
I wonder if, in a new target, we could add elements to those lists that contain the `model_name` and `out_dir` and then map over each list to populate the arguments in `plot_vip`. I don't think I've tried mapping over a nested list, but I don't see why it wouldn't work...

jds485 replied:
That sounds good! I can add that idea to the enhancement issue

lekoenig reviewed (approved):
Thanks @jds485, these plots are looking great and will be so helpful for diagnosing and interpreting model outputs! 🎉

I've gone ahead and "approved" this PR because I view all of my comments as optional suggestions that you can consider if time allows.

jds485 commented on 3_visualize/src/plot_nhdv2_attr.R:
```diff
@@ -30,11 +33,17 @@ plot_nhdv2_attr <- function(attr_data,network_geometry,file_path){
+    attr_plot_spatial <- ggplot() + 
+      #full network (in case there are reaches without data)
+      geom_sf(data = network_geometry, 
+              size = 0.3, color = 'gray') +
```
Yes, that's right

jds485 commented on 4_predict/src/plot_diagnostics.R:
```diff
@@ -0,0 +1,559 @@
+  
+  p1 <- opt_result %>% 
+    tune::collect_metrics() %>%
+    ggplot(aes(mtry, mean, color = min_n)) +
```
it's a hyperparameter for RF that specifies the number of attributes to randomly sample at each split in the tree. I can add a comment about that and min_n

lekoenig replied:
If they are short comments to explain the plot, you could even add them to the plot as a subtitle, e.g. `ggtitle(label = model_name, subtitle = "mtry is the number of attr to randomly sample at each split in the tree"`.

jds485 commented on 4_predict/src/plot_diagnostics.R:
```diff
@@ -0,0 +1,559 @@
+  #'
+  #' @return filepath to the resulting plot
+  
+  if(from_predict){
+    #predict from provided workflow and data
+    df_pred_obs <- predict(model_wf, new_data = pred_data, type = 'numeric') %>%
+      mutate(obs = pred_data[[pred_var]])
+  }
```
Not currently. I copied most of this function from the FHWA repo where we use the `from_predict` capability instead of making a target containing the new predictions in an object like `df_pred_obs`. In this repo I've created targets with the predictions because the predictions are used for other plots and targets

lekoenig replied:
Ah, OK. I generally like the idea of having the plotting function stick to just creating plots, but I'll keep this in mind.

jds485 commented on 4_predict/src/plot_diagnostics.R:
```diff
@@ -0,0 +1,559 @@
+  if(count_shade){
+    fileout <- paste0('pred_obs_scatter_', model_name, '_density.png')
+    
+    p1 <- ggplot(df_pred_obs, aes(x=log10(obs), y=log10(.pred))) +
```
Yes, `.pred` is a column name. I'll add a comment - that is confusing

lekoenig replied:
OK, thanks for clarifying! tidyverse changes syntax styles often enough that I just assumed you had a new trick here 🙂

jds485 commented on 4_predict/src/plot_diagnostics.R:
```diff
@@ -0,0 +1,559 @@
+      theme_bw() +
+      theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
+      xlim(0,5) + ylim(0,5) +
+      xlab('Observed') + ylab('Predicted') +
```
Good point! I think we can add hard-coded units for now and revise later if we predict a different constituent.

jds485 commented on 4_predict/src/plot_diagnostics.R:
```diff
@@ -0,0 +1,559 @@
+                  position = position_dodge(0.6)) +
+    xlab('') +
+    ylab(perf_metric) + 
+    scale_x_discrete(limits=c("RF-Static")) +
```
Thanks! That should be the `model_name` function arg

jds485 commented on 4_predict/src/plot_diagnostics.R:
```diff
@@ -0,0 +1,559 @@
+    geom_bar(stat="identity", position=position_dodge(), width = 0.6) +
+    theme_bw() +
+    scale_fill_brewer(palette="Paired") +
+    geom_errorbar(aes(ymin = perf - 2*sd, ymax = perf + 2*sd), width = .2,
```
Yes, I currently do not have multiple test evaluations. I have several validation sets from the cross validation, but only 1 test set.

jds485 commented on 4_predict/src/plot_diagnostics.R:
```diff
@@ -0,0 +1,559 @@
+  fileout <- file.path(out_dir, 
+                       paste0('SHAP_global_', model_name, '.png'))
+  
+  p1 <- autoplot(shap, num_features = num_features) +
```
It seems like a wrapper around ggplot, and packages develop their own methods. This one is from fastshap, so I'll add `fastshap::`.

jds485 commented on 4_predict/src/plot_diagnostics.R:
```diff
@@ -0,0 +1,559 @@
+  return(fileout)
+}
+
+plot_shap_dependence <- function(shap, data, model_name, out_dir, ncores = 1){
```
I must not have built these on Tallgrass. I can upload the plots I created locally so you can see what these are like.

jds485 replied:
Uploading for `4_predict/out/temporal/shap/RF_min_static_dynamic/`

jds485 commented on _targets.R:
```diff
@@ -15,7 +15,9 @@ tar_option_set(packages = c("tidyverse", "lubridate",
-                            "patchwork", "glue",'nhdplusTools'),
+                            "patchwork", "glue", "nhdplusTools",
+                            "Boruta", "ranger", "vip", "tidymodels",
+                            "doParallel", "fastshap", "pdp"),
```
`tune` is part of `tidymodels` and `vip` is included

jds485 commented on 4_predict/src/plot_diagnostics.R:
```diff
@@ -0,0 +1,559 @@
+  plt_lst <- list()
+  
+  # For each column, create a barplot
+  for(i in 2:dim(attr_data)[2]){
```
I think I'll leave this function as-is for now because I'd need to add another argument to specify the x-axis column name to use `seq_along` for the y-axes. I did implement it for the nhdv2 attr plotting, though.

galengorski reviewed (approved):
@jds485 Looks good and got the target `p4_train_RF_min_static` to run on tallgrass as a spot check for running the models. One thing that we discussed in troubleshooting, was to include instructions for how to review a PR for this pipeline within the container on tallgrass in the readme file. Other than that, I had a few questions and points of clarification.

jds485 commented on 4_predict/src/select_features.R:
```diff
@@ -0,0 +1,273 @@
+  #' highly correlated with other attributes. 
+  #'
+  #' @param features table of features (columns) for each prediction point (rows).
+  #' @param drop_columns character vector of column names to remove from features. 
+  #' This is passed to dplyr::contains()
+  #' @param drop_exact character vector of exact column names to drop.
+  #' This is passed to dplyr::select()
```
The difference is that `drop_exact` uses `select()` to remove the column, which requires the exact column name, and `drop_columns` uses `contains()` to remove the column, which matches all column names with a particular pattern. I can rename these args to be more helpful, though:
`drop_columns_select`, and `drop_columns_contains`

lekoenig replied:
I see the difference between these options, but agree with Galen that having both options potentially makes the `drop_high_corr_attrs()` function harder to grasp. You could consider just using `drop_columns_contains` and then an exact string should still work with `dplyr::contains()`. I don't have a strong opinion on that, but this thread reminds me that I've sometimes been surprised by unexpected string matches when typing out options like those currently included in `drop_columns`. It might be a nice feature to print a message that confirms which columns are dropped as long as the output isn't so verbose that a user is likely to just ignore it altogether (here's an [example](https://github.com/USGS-R/ds-pipelines-targets-example-wqp/blob/main/1_inventory/src/check_characteristics.R#L77-L79) of that idea, but with a warning message).

lekoenig replied:
Thanks, @jds485 - that's a good example of this sort of edge case.

jds485 commented on 4_predict/src/train_models.R:
```diff
@@ -0,0 +1,500 @@
+  return(brf_output)
+}
+
+make_temporal_split <- function(attrs_df, train_prop){
+  #' 
```
`make_temporal_split` calls `split_data` to make a temporal split. `make_temporal_split` has a slightly different input data structure

jds485 commented on 4_predict.R:
```diff
@@ -0,0 +1,589 @@
+  #            },
+  #            deployment = 'worker'
+  # ),
+  tar_target(p4_train_RF_static_dynamic,
+             {#Filter out data before 1984-09-30 for training due to NAs
```
I do have comment descriptions before each of the `p4_train...` targets, but the commented out code may have made it easy to miss:
```
#dynamic only
  # tar_target(p4_train_RF_dynamic,
  #            train_models_grid(brf_output = p4_dynamic_attrs,
  #                              ncores = RF_cores,
  #                              v_folds = cv_folds,
  #                              range_mtry = c(5,30),
  #                              range_minn = c(2,20),
  #                              range_trees = c(100,500),
  #                              gridsize = 50,
  #                              id_cols = c('PRMS_segid', 'Date')
  #            ),
  #            deployment = 'worker'
  # ),
  ```

galengorski replied:
Oh I see how it is organized now, ok that makes sense

jds485 commented on 4_predict/src/select_features.R:
```diff
@@ -0,0 +1,273 @@
+  }
+  
+  #Drop rows with NAs
+  features <- drop_na(features)
```
> Does this drop the entire row if a feature is NA?

Yes, it drops the whole row. In this case, that amounts to dropping data before 1985 water year. 
```
max(features[which(is.na(features), arr.ind = T)[,1], c(1,unique(which(is.na(features), arr.ind = T)[,2]))]$Date)
[1] "1984-09-29"
```
I kept it general in this function in case new attributes are added that have NAs that are not handled before this function.

> What if you end up not using the feature that yields the NA value?

For this function, the correlation is only checked for rows with all feature data available. For modeling, rows with NAs are still available to use if the features are not selected for use. This function returns the column names of features to use.

jds485 commented:
Thanks @lekoenig and @galengorski! I think I addressed all of your comments. I'm now updating the targets and will commit the meta file before merging
